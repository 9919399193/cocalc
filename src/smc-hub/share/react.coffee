###
Load react support for rendering.
###

async = require('async')

# Code for rendering react components to html.
ReactDOMServer = require('react-dom/server')

require('./jsdom-support')

# Uncomment for cc-in-cc dev benchmarking purposes.  This variable is already set
# by the Docker container when running in kubernetes.
## process.NODE_ENV="production"

require('smc-webapp/r_misc').SHARE_SERVER = true

# Load katex jQuery plugin.
require('smc-webapp/jquery-plugins/katex')

exports.react = (res, component, extra, viewer) ->
    res.type('html')
    t0 = new Date()
    stream = ReactDOMServer.renderToStaticNodeStream(component)
    stream.pipe(res)
    stream.once 'end', ->
        console.log("react: time to render and stream out: #{new Date() - t0}ms", extra)
