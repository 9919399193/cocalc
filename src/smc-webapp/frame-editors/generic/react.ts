// THIS SHOULD BE MOVED OUT OF frame-editors/


export const {
  React,
  ReactDOM,
  rclass,
  rtypes,
  Fragment,
  redux,
  Redux
} = require("smc-webapp/smc-react");

import * as ReactOrig from "react";

export const Component = ReactOrig.Component;

export type Rendered = ReactOrig.ReactElement<any> | undefined;
