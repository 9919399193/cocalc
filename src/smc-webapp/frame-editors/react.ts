export const {
  React,
  ReactDOM,
  rclass,
  rtypes,
  Fragment,
  redux,
  Redux
} = require("smc-webapp/smc-react");

import * as ReactOrig from "react";

export const Component = ReactOrig.Component;

export type Rendered = ReactOrig.ReactElement<any> | undefined;
