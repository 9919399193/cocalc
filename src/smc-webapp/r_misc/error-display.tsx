import * as React from "react";
import * as misc from "smc-util/misc";
import { Alert } from "react-bootstrap";
import { CloseX } from "./close-x";

const error_text_style = {
  marginRight: "1ex",
  whiteSpace: "pre-line",
  maxWidth: "80ex"
};

interface Props {
  error?: string | object;
  error_component?: React.ComponentType;
  title?: string;
  style?: React.CSSProperties;
  bsStyle?: string;
  onClose?: () => void;
}

export class ErrorDisplay extends React.Component<Props> {
  render_close_button() {
    if (this.props.onClose == undefined) {
      return;
    }
    return (
      <CloseX on_close={this.props.onClose} style={{ fontSize: "11pt" }} />
    );
  }

  render_title() {
    return <h4>{this.props.title}</h4>;
  }

  render() {
    let error, style;

    if (this.props.style != undefined) {
      style = misc.copy(error_text_style);
      misc.merge(style, this.props.style);
    } else {
      style = error_text_style;
    }

    if (this.props.error != undefined) {
      if (typeof this.props.error === "string") {
        error = this.props.error;
      } else {
        error = misc.to_json(this.props.error);
      }
    } else {
      error = this.props.error_component;
    }

    const bsStyle = this.props.bsStyle != undefined ? this.props.bsStyle : "danger";

    return (
      <Alert bsStyle={bsStyle} style={style}>
        {this.render_close_button()}
        {this.props.title ? this.render_title() : undefined}
        {error}
      </Alert>
    );
  }
}
