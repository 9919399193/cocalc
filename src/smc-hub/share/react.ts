/*
Load react support for rendering.
*/

// Code for rendering react components to html.
import * as ReactDOMServer from "react-dom/server";

import "./jsdom-support";

// Uncomment for cc-in-cc dev benchmarking purposes.  This variable is already set
// by the Docker container when running in kubernetes.
//# process.NODE_ENV="production"

import { set_share_server } from "smc-webapp/r_misc/share-server";

set_share_server(true);

// Load katex jQuery plugin.
require("smc-webapp/jquery-plugins/katex");

export function react(res, component, extra: any = ""): void {
  res.type("html");
  const t0 = new Date().valueOf();
  const stream = ReactDOMServer.renderToStaticNodeStream(component);
  stream.pipe(res);
  stream.once("end", () =>
    console.log(
      `react: time to render and stream out: ${new Date().valueOf() - t0}ms`,
      extra
    )
  );
}
