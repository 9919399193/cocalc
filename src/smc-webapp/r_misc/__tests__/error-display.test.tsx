import * as React from "react";
import { shallow } from "enzyme";
import { ErrorDisplay } from "../error-display";

test("smoke test", () => {
  const rendered = shallow(
    <ErrorDisplay error={"Testing error"} onClose={() => undefined} />
  );
  expect(rendered).toMatchSnapshot();
});
