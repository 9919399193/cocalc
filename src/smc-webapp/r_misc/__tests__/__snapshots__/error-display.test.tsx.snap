// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke test 1`] = `
<Alert
  bsClass="alert"
  bsStyle="danger"
  closeLabel="Close alert"
  style={
    Object {
      "marginRight": "1ex",
      "maxWidth": "80ex",
      "whiteSpace": "pre-line",
    }
  }
>
  <CloseX
    on_close={[Function]}
    style={
      Object {
        "fontSize": "11pt",
      }
    }
  />
  Testing error
</Alert>
`;
